src/main/java/com/artipie/metrics/prometheus/PushGateway.java

```diff
@@ -0,0 +1,301 @@
+package com.artipie.metrics.prometheus;
+
+import java.io.BufferedWriter;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStreamWriter;
+import java.io.UnsupportedEncodingException;
+import java.net.HttpURLConnection;
+import java.net.InetAddress;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URL;
+import java.net.URLEncoder;
+import java.net.UnknownHostException;
+import java.util.HashMap;
+import java.util.Map;
+import javax.xml.bind.DatatypeConverter;
+
+import io.prometheus.client.Collector;
+import io.prometheus.client.CollectorRegistry;
+import io.prometheus.client.exporter.common.TextFormat;
+
+/**
+ * Export metrics via the Prometheus Pushgateway.
+ * <p>
+ * The Prometheus Pushgateway exists to allow ephemeral and batch jobs to expose their metrics to Prometheus.
+ * Since these kinds of jobs may not exist long enough to be scraped, they can instead push their metrics
+ * to a Pushgateway. This class allows pushing the contents of a {@link CollectorRegistry} to
+ * a Pushgateway.
+ * <p>
+ * Example usage:
+ * <pre>
+ * {@code
+ *   void executeBatchJob() throws Exception {
+ *     CollectorRegistry registry = new CollectorRegistry();
+ *     Gauge duration = Gauge.build()
+ *         .name("my_batch_job_duration_seconds").help("Duration of my batch job in seconds.").register(registry);
+ *     Gauge.Timer durationTimer = duration.startTimer();
+ *     try {
+ *       // Your code here.
+ *
+ *       // This is only added to the registry after success,
+ *       // so that a previous success in the Pushgateway isn't overwritten on failure.
+ *       Gauge lastSuccess = Gauge.build()
+ *           .name("my_batch_job_last_success").help("Last time my batch job succeeded, in unixtime.").register(registry);
+ *       lastSuccess.setToCurrentTime();
+ *     } finally {
+ *       durationTimer.setDuration();
+ *       PushGateway pg = new PushGateway("127.0.0.1:9091");
+ *       pg.pushAdd(registry, "my_batch_job");
+ *     }
+ *   }
+ * }
+ * </pre>
+ * <p>
+ * See <a href="https://github.com/prometheus/pushgateway">https://github.com/prometheus/pushgateway</a>
+ */
+public class PushGateway {
+
+    private static final int MILLISECONDS_PER_SECOND = 1000;
+
+    // Visible for testing.
+    protected final String gatewayBaseURL;
+
+    private HttpConnectionFactory connectionFactory = new DefaultHttpConnectionFactory();
```

Comment:
@milkysoft looks like this should be a final field initialized in the ctor, also compound names are not allowed by our checkstyle. Please, read this [blog post](https://www.yegor256.com/2015/01/12/compound-name-is-code-smell.html).